pkg/maps/metricsmap/metricsmap.go

```diff
@@ -183,23 +177,88 @@ func SyncMetricsMap() error {
 
 	var key, nextKey Key
 	for {
-		err := metricsmap.GetNextKey(&key, &nextKey)
+		err := bpf.GetNextKey(metricsmap.GetFd(), unsafe.Pointer(&key), unsafe.Pointer(&nextKey))
 		if err != nil {
 			break
 		}
-		entry, err := metricsmap.Lookup(&nextKey)
+		err = bpf.LookupElement(metricsmap.GetFd(), unsafe.Pointer(&nextKey), unsafe.Pointer(&entry))
 		if err != nil {
 			return fmt.Errorf("unable to lookup metrics map: %s", err)
 		}
-		value := entry.(*Value)
-		// Increment Prometheus metrics here.
-		updatePrometheusMetrics(&nextKey, value)
-		key = nextKey
+
+		// cannot use `range entry` since, if the first value for a particular
+		// CPU is zero, it never iterates over the next non-zero value.
+		for i := 0; i < possibleCpus; i++ {
+			// Increment Prometheus metrics here.
+			updatePrometheusMetrics(&nextKey, &entry[i])
+			key = nextKey
+		}
+
 	}
 	return nil
 }
 
+// calculateNumCpus replicates the bpf linux helper equivalent `bpf_num_possible_cpus`
+// to find total number of possible CPUs i.e cpus that have been allocated
+// resources and can be brought online if they are present.
+// Reference bpf_num_possible_cpus from
+// https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf.git/tree/tools/testing/selftests/bpf/bpf_util.h
+func calculateNumCpus() {
+	var start, end int
+
+	file, err := os.Open("/sys/devices/system/cpu/possible")
+	if err != nil {
+		log.WithError(err).Error("unable to open sysfs to get CPU count")
+	}
+	defer file.Close()
+
+	data := make([]byte, maxCPUS)
+	for {
+		_, err := file.Read(data)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			log.WithError(err).Error("unable to read sysfs to get CPU count")
+		}
+		n, err := fmt.Sscanf(string(data), "%d-%d", &start, &end)
+		if err != nil {
+			log.WithError(err).Error("unable to parse sysfs to get CPU count")
+		}
+		if n == 0 {
+			log.WithError(err).Error("Failed to retrieve # possible CPUs!")
+		} else if n == 1 {
+			end = start
+		}
+		if start == 0 {
+			possibleCpus = end + 1
+		} else {
+			possibleCpus = 0
+		}
+		break
+	}
+}
+
 func init() {
+	calculateNumCpus()
+	// Metrics is a mapping of all packet drops and forwards associated with
+	// the node on ingress/egress direction
+	Metrics = bpf.NewMap(
+		MapName,
+		bpf.BPF_MAP_TYPE_PERCPU_HASH,
```

Comment:
Is there any concern with supported kernel versions with using this map type given that we haven't used it before AFAICT? 
cc: @joestringer  - are there any additional tests that may need to be performed / are recommended to ensure that this map type has all of its functions implemented? I mention this due to the issues we have seen with incomplete implementation of LPM in some kernel versions that were only implemented in later versions of the kernel.

Reply:
It does look like it was introduced in 4.9 @borkmann : Should be be concerned about breaking any specific linux versions?

Reply:
@ianvernon I think that if it's tested in an environment using the ginkgo VM and its v4.9 kernel, observe the traffic flow through and the metrics get updated and make sure the logs don't have any weird messages then that's probably enough. @borkmann would have a better idea if there's something that may be missed, though.